content/post/2023-01-24-code-switch-escape-hatch-test/index.Rmd

```diff
@@ -0,0 +1,116 @@
+date: "2023-01-24" 
+tags: 
+- package development 
+output: hugodown::hugo_document
+---
+
+```{r setup, include=FALSE}
+
+knitr::opts_chunk$set(fig.path = "", comment = "")
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+Sometimes, testing [gets hard](https://r-pkgs.org/testing-advanced.html#when-testing-gets-hard).
+For instance, you'd like to test for the behavior of your function in the absence of an internet connection, or in an interactive session, without actually cutting off the internet, or from the safety of a definitely non interactive R session for tests.
+In this post we shall present a not too involved pattern to avoid very complicated infrastructure, as a complement to [mocking](/2019/10/29/mocking/) in your toolbelt.
+
+## The pattern
+
+Say my package code displays a message "No internet! Le sigh" when there's no internet, and I want to test for that message.
+
+First, I create a function called `is_internet_down()`.
+It could simply call `curl::has_internet()`.
+I will use it from my code.
+
+``` r
+is_internet_down <- function() {
+  curl::has_internet()
+}
+
+my_complicated_code <- function() {
+  if (is_internet_down()) {
+    message("No internet! Le sigh")
+  }
+  # blablablabla
+}
+```
+
+Now in tests, I can't catch the message if there is internet.
+
+``` r
+test_that("my_complicated_code() notes the absence of internet", {
+  expect_message(my_complicated_code(), "No internet")
+})
+```
+
+This is where I add a switch to my code!
+
+``` r
+is_internet_down <- function() {
+
+  if (nzchar(Sys.getenv("TESTPKG.NOINTERNET"))) {
+    return(TRUE)
+  }
+
+  curl::has_internet()
+}
+
+my_complicated_code <- function() {
+  if (is_internet_down()) {
+    message("No internet! Le sigh")
+  }
+  # blablablabla
+}
+```
+
+Now, when the environment variable "TESTPKG.NOINTERNET" is set to something, anything, my function `is_internet_down()` will return `TRUE` and my code will show the message.
+Note that I tried to name the code switch to something readable.
+
+In the tests, I add a call to withr to set that environment variable for the test only.
+
+``` r
+test_that("my_complicated_code() notes the absence of internet", {
+  withr::local_envvar("TESTPKG.NOINTERNET" = "blop")
+  expect_message(my_complicated_code(), "No internet")
+})
+```
+
+That's all there is to the pattern.
+You could use an option and `withr::local_options()` instead.
+
+## Use of the pattern in the wild
+
+A popular example of a function with an [escape hatch](https://twitter.com/JennyBryan/status/1613976157501927424) is [`rlang::is_interactive()`](https://rlang.r-lib.org/reference/is_interactive.html).
```

Comment:
Would it make sense to embed the tweet to avoid recent issues where tweets have been deleted?

Reply:
Good point. I've phrased the sentence so that removing the link wouldn't be too problematic, especially as the "escape hatch" phrase is in rlang docs (but I hadn't noticed it prior to the tweet). Does that make sense?